// @lingui/cli/dist/services/translationIO/segment-converters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSegmentFromLinguiItem = createSegmentFromLinguiItem;
exports.createLinguiItemFromSegment = createLinguiItemFromSegment;
const generateMessageId_1 = require("@lingui/message-utils/generateMessageId");
const EXPLICIT_ID_FLAG = "js-lingui-explicit-id";
const EXPLICIT_ID_AND_CONTEXT_FLAG = "js-lingui-explicit-id-and-context";
function isGeneratedId(id, message) {
    return id === (0, generateMessageId_1.generateMessageId)(message.message, message.context);
}
const joinOrigin = (origin) => origin.join(":");
const splitOrigin = (origin) => {
    const [file, line] = origin.split(":");
    return [file, line ? Number(line) : null];
};
function createSegmentFromLinguiItem(key, item) {
    const itemHasExplicitId = !isGeneratedId(key, item);
    const itemHasContext = !!item.context;
    const segment = {
        type: "source", // No way to edit text for source language (inside code), so not using "key" here
        source: "",
        context: "",
        references: [],
        comment: "",
    };
    // For segment.source & segment.context, we must remain compatible with projects created/synced before Lingui V4
    if (itemHasExplicitId) {
        segment.source = item.message || item.translation;
        segment.context = key;
    }
    else {
        segment.source = item.message || item.translation;
        if (itemHasContext) {
            segment.context = item.context;
        }
    }
    if (item.origin) {
        segment.references = item.origin.map(joinOrigin);
    }
    // Since Lingui v4, when using explicit IDs, Lingui automatically adds 'js-lingui-explicit-id' to the extractedComments array
    const comments = [];
    if (itemHasExplicitId) {
        if (itemHasContext) {
            // segment.context is already used for the explicit ID, so we need to pass the context (for translators) in segment.comment
            comments.push(item.context, EXPLICIT_ID_AND_CONTEXT_FLAG);
        }
        else {
            comments.push(EXPLICIT_ID_FLAG);
        }
    }
    segment.comment = [...comments, ...(item.comments || [])].join(" | ");
    return segment;
}
function createLinguiItemFromSegment(segment) {
    var _a, _b, _c;
    const segmentHasExplicitId = (_a = segment.comment) === null || _a === void 0 ? void 0 : _a.includes(EXPLICIT_ID_FLAG);
    const segmentHasExplicitIdAndContext = (_b = segment.comment) === null || _b === void 0 ? void 0 : _b.includes(EXPLICIT_ID_AND_CONTEXT_FLAG);
    const item = {
        translation: segment.target,
        origin: ((_c = segment.references) === null || _c === void 0 ? void 0 : _c.length)
            ? segment.references.map(splitOrigin)
            : [],
        message: segment.source,
        comments: [],
    };
    let id = null;
    if (segmentHasExplicitId || segmentHasExplicitIdAndContext) {
        id = segment.context;
    }
    else {
        id = (0, generateMessageId_1.generateMessageId)(segment.source, segment.context);
        item.context = segment.context;
    }
    if (segment.comment) {
        item.comments = segment.comment.split(" | ").filter(
        // drop flags from comments
        (comment) => comment !== EXPLICIT_ID_AND_CONTEXT_FLAG && comment !== EXPLICIT_ID_FLAG);
        // We recompose a target PO Item that is consistent with the source PO Item
        if (segmentHasExplicitIdAndContext) {
            item.context = item.comments.shift();
        }
    }
    return [id, item];
}
