// @lingui/cli/dist/services/translationIO/translationio-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tioSync = tioSync;
exports.tioInit = tioInit;
// todo: need to enable strictNullChecks to support this kind of type narrowing
// export type TranslationIoResponse =
//   | TranslationIoErrorResponse
//   | TranslationProjectResponse
//
// type TranslationIoErrorResponse = {
//   errors: string[]
// }
// type TranslationProjectResponse = {
//   errors: null
//   project: TranslationIoProject
//   segments: { [locale: string]: TranslationIoSegment[] }
// }
async function post(url, request) {
    const response = await fetch(url, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(request),
    });
    if (!response.ok && response.status !== 400) {
        return {
            error: {
                response,
                message: `Request failed with ${response.status} ${response.statusText} status. Body: ${await response.text()}`,
            },
            data: undefined,
        };
    }
    return { data: await response.json(), error: undefined };
}
async function tioSync(request, apiKey) {
    return post(`https://translation.io/api/v1/segments/sync.json?api_key=${apiKey}`, request);
}
async function tioInit(request, apiKey) {
    return post(`https://translation.io/api/v1/segments/init.json?api_key=${apiKey}`, request);
}
