// @lingui/cli/dist/services/translationIO.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = syncProcess;
exports.init = init;
exports.sync = sync;
exports.writeSegmentsToCatalogs = writeSegmentsToCatalogs;
const fs_1 = __importDefault(require("fs"));
const path_1 = require("path");
const translationio_api_1 = require("./translationIO/translationio-api");
const catalog_1 = require("../api/catalog");
const segment_converters_1 = require("./translationIO/segment-converters");
const getTargetLocales = (config) => {
    const sourceLocale = config.sourceLocale || "en";
    const pseudoLocale = config.pseudoLocale || "pseudo";
    return config.locales.filter((value) => value != sourceLocale && value != pseudoLocale);
};
// Main sync method, call "Init" or "Sync" depending on the project context
async function syncProcess(config, options, extractionResult) {
    const reportSuccess = (project) => {
        return `\n----------\nProject successfully synchronized. Please use this URL to translate: ${project.url}\n----------`;
    };
    const reportError = (errors) => {
        throw `\n----------\nSynchronization with Translation.io failed: ${errors.join(", ")}\n----------`;
    };
    const { success, project, errors } = await init(config, extractionResult);
    if (success) {
        return reportSuccess(project);
    }
    if (errors[0] === "This project has already been initialized.") {
        const { success, project, errors } = await sync(config, options, extractionResult);
        if (success) {
            return reportSuccess(project);
        }
        return reportError(errors);
    }
    return reportError(errors);
}
// Initialize project with source and existing translations (only first time!)
// Cf. https://translation.io/docs/create-library#initialization
async function init(config, extractionResult) {
    const sourceLocale = config.sourceLocale || "en";
    const targetLocales = getTargetLocales(config);
    const segments = {};
    targetLocales.forEach((targetLocale) => {
        segments[targetLocale] = [];
    });
    // Create segments from source locale PO items
    for (const { messagesByLocale } of extractionResult) {
        const messages = messagesByLocale[sourceLocale];
        Object.entries(messages).forEach(([key, entry]) => {
            if (entry.obsolete)
                return;
            targetLocales.forEach((targetLocale) => {
                segments[targetLocale].push((0, segment_converters_1.createSegmentFromLinguiItem)(key, entry));
            });
        });
    }
    // Add translations to segments from target locale PO items
    for (const { messagesByLocale } of extractionResult) {
        for (const targetLocale of targetLocales) {
            const messages = messagesByLocale[targetLocale];
            Object.entries(messages)
                .filter(([, entry]) => !entry.obsolete)
                .forEach(([, entry], index) => {
                segments[targetLocale][index].target = entry.translation;
            });
        }
    }
    const { data, error } = await (0, translationio_api_1.tioInit)({
        client: "lingui",
        version: require("@lingui/core/package.json").version,
        source_language: sourceLocale,
        target_languages: targetLocales,
        segments: segments,
    }, config.service.apiKey);
    if (error) {
        return { success: false, errors: [error.message] };
    }
    if (data.errors) {
        return { success: false, errors: data.errors };
    }
    await writeSegmentsToCatalogs(config, sourceLocale, extractionResult, data.segments);
    return { success: true, project: data.project };
}
// Send all source text from PO to Translation.io and create new PO based on received translations
// Cf. https://translation.io/docs/create-library#synchronization
async function sync(config, options, extractionResult) {
    const sourceLocale = config.sourceLocale || "en";
    const targetLocales = getTargetLocales(config);
    const segments = [];
    // Create segments with correct source
    for (const { messagesByLocale } of extractionResult) {
        const messages = messagesByLocale[sourceLocale];
        Object.entries(messages).forEach(([key, entry]) => {
            if (entry.obsolete)
                return;
            segments.push((0, segment_converters_1.createSegmentFromLinguiItem)(key, entry));
        });
    }
    const { data, error } = await (0, translationio_api_1.tioSync)({
        client: "lingui",
        version: require("@lingui/core/package.json").version,
        source_language: sourceLocale,
        target_languages: targetLocales,
        segments: segments,
        // Sync and then remove unused segments (not present in the local application) from Translation.io
        purge: Boolean(options.clean),
    }, config.service.apiKey);
    if (error) {
        return { success: false, errors: [error.message] };
    }
    if (data.errors) {
        return { success: false, errors: data.errors };
    }
    await writeSegmentsToCatalogs(config, sourceLocale, extractionResult, data.segments);
    return { success: true, project: data.project };
}
async function writeSegmentsToCatalogs(config, sourceLocale, extractionResult, segmentsPerLocale) {
    // Create segments from source locale PO items
    for (const { catalog, messagesByLocale } of extractionResult) {
        const sourceMessages = messagesByLocale[sourceLocale];
        for (const targetLocale of Object.keys(segmentsPerLocale)) {
            // Remove existing target POs and JS for this target locale
            {
                const path = catalog.getFilename(targetLocale);
                const jsPath = path.replace(new RegExp(`${catalog.format.getCatalogExtension()}$`), "") + ".js";
                const dirPath = (0, path_1.dirname)(path);
                // todo: check tests and all these logic, maybe it could be simplified to just drop the folder
                // Remove PO, JS and empty dir
                if (fs_1.default.existsSync(path)) {
                    await fs_1.default.promises.unlink(path);
                }
                if (fs_1.default.existsSync(jsPath)) {
                    await fs_1.default.promises.unlink(jsPath);
                }
                if (fs_1.default.existsSync(dirPath) && fs_1.default.readdirSync(dirPath).length === 0) {
                    await fs_1.default.promises.rmdir(dirPath);
                }
            }
            const translations = Object.fromEntries(segmentsPerLocale[targetLocale].map((segment) => (0, segment_converters_1.createLinguiItemFromSegment)(segment)));
            const messages = Object.fromEntries(Object.entries(sourceMessages).map(([key, entry]) => {
                var _a;
                return [
                    key,
                    Object.assign(Object.assign({}, entry), { translation: (_a = translations[key]) === null || _a === void 0 ? void 0 : _a.translation }),
                ];
            }));
            await catalog.write(targetLocale, (0, catalog_1.order)(config.orderBy, messages));
        }
    }
}
